src/librustc/ty/sty.rs

```diff
@@ -738,8 +733,16 @@ impl<'a, 'tcx, 'gcx> PolyExistentialProjection<'tcx> {
         self.0.item_name // safe to skip the binder to access a name
     }
 
-    pub fn sort_key(&self) -> (DefId, Name) {
-        (self.0.trait_ref.def_id, self.0.item_name)
+    pub fn sort_key(&self, hasher: &mut DefPathHashes) -> (u64, InternedString) {
+        // We want something here that is stable across trait boundaries.
+        // The DefId isn't but the `deterministic_hash` of the corresponding
+        // DefPath is.
+        let def_path_hash = hasher.hash(self.0.trait_ref.def_id);
```

Comment:
I think so, it's just a hash of crate name + disambiguator and the DefPath components.